client/src/main/java/io/pravega/client/segment/impl/SegmentOutputStreamImpl.java

```diff
@@ -354,68 +341,73 @@ public void processingFailure(Exception error) {
-                getConnection(); // As the messages is inflight, this will perform the retransmition.
+                try {
+                    getConnection(); // As the messages is inflight, this will perform the retransmition.
+                } catch (SegmentSealedException e2) {
+                    return;
+                }
             }
         }
     }
     
     /**
-     * Blocking call to establish a connection and wait for it to be setup. (Retries built in)
+     * Establish a connection and wait for it to be setup. (Retries built in)
      */
-    ClientConnection getConnection() throws SegmentSealedException {
+    CompletableFuture<ClientConnection> getConnection() throws SegmentSealedException {
         checkState(!state.isClosed(), "LogOutputStream was already closed");
         if (state.isAlreadySealed()) {
             throw new SegmentSealedException(this.segmentName);
         }
-        return retrySchedule.retryingOn(ConnectionFailedException.class).throwingOn(SegmentSealedException.class).run(() -> {
-            setupConnection();
-            return state.waitForConnection();
-        });
+        return retrySchedule.retryingOn(ConnectionFailedException.class).throwingOn(SegmentSealedException.class).runAsync(() -> {
+            return initiateConnectionSetup();
+        }, connectionFactory.getInternalExecutor());
     }
-
+    
     @VisibleForTesting
-    void setupConnection() throws ConnectionFailedException {
-        if (state.getConnection() == null) {
-            synchronized (connectionEstablishmentLock) {
-                if (state.getConnection() == null && !state.isAlreadySealed()) {
-                    log.info("Fetching endpoint for segment {}", segmentName);
-                    CompletableFuture<ClientConnection> newConnection = controller.getEndpointForSegment(segmentName)
-                            .thenCompose((PravegaNodeUri uri) -> {
-                                log.info("Establishing connection to {} for {}", uri, segmentName);
-                                return connectionFactory.establishConnection(uri, responseProcessor);
-                            });
-                    ClientConnection connection = getAndHandleExceptions(newConnection, ConnectionFailedException::new);
-                    state.newConnection(connection);
-                    SetupAppend cmd = new SetupAppend(requestIdGenerator.get(), writerId, segmentName);
-                    try {
-                        connection.send(cmd);
-                    } catch (Exception e) {
-                        state.failConnection(e);
-                        throw e;
-                    }
-                }
+    CompletableFuture<ClientConnection> initiateConnectionSetup() {
+        CompletableFuture<ClientConnection> future =  new CompletableFuture<>();
+        state.connectionSetup.runReleaserAndAwait(this::setupConnection, future);
```

Comment:
`setupConnection()` throwing a runtime exception will cause the method `runReleaserAndAwait` to thrown a run time exception too(i.e. the future will not be completed.). 
this will also result in getConnection() method throwing this runtimeException. i.e. its retry is not exercised, since it is not of type `ConnectionFailedException`.
Also, we retry indefinitely incase of a similar exception inside `io.pravega.client.segment.impl.SegmentOutputStreamImpl#failConnection`